clients/src/main/java/org/apache/kafka/common/protocol/types/Type.java

```diff
@@ -184,14 +184,19 @@ public Long validate(Object item) {
         public void write(ByteBuffer buffer, Object o) {
             byte[] bytes = Utils.utf8((String) o);
             if (bytes.length > Short.MAX_VALUE)
-                throw new SchemaException("String is longer than the maximum string length.");
+                throw new SchemaException("String length " + bytes.length + " is larger than the maximum string length.");
             buffer.putShort((short) bytes.length);
             buffer.put(bytes);
         }
 
         @Override
         public Object read(ByteBuffer buffer) {
-            int length = buffer.getShort();
+            short length = buffer.getShort();
+            if (length < 0)
+                throw new SchemaException("String length " + length + " cannot be negative");
+            if (length > buffer.remaining())
```

Comment:
If the goal is preventing generic "BufferUnderflowExceptions" I think its worth adding. Doesn't need to be this patch if you don't want to though.